Fix 'particle density' mesh refinement

gassmoeller opened the pull request:
The refinement criterion 'particle density' did not decrease when the cells were refined, which was not caught by the test, because it only refined once. The correct measure is the number of particles per cell (equivalent to: particle_density * cell_volume). The new test captures this behavior.

bangerth commented:
There is one test that fails on the tester. Any insight?

bangerth commented on source/mesh_refinement/particle_density.cc:
```diff
@@ -35,7 +35,7 @@ namespace aspect
 
       AssertThrow(tracer_postprocessor != 0,
-                  ExcMessage("The mesh refinement plugin 'tracer distribution' requires the "
+                  ExcMessage("The mesh refinement plugin 'particle density' requires the "
```
good catch. that would have been one awkward to figure out error message...

bangerth commented on source/mesh_refinement/particle_density.cc:
```diff
@@ -54,7 +54,7 @@ namespace aspect
             // This is unaccurate for curved cells, but the exact measure is likely not
             // important for this plugin since the cell distortion is usually small.
-            indicators(i) = n_tracers / cell->measure();
+            indicators(i) = static_cast<float>(n_tracers);
```
Hm, I don't think I quite understand what you're doing here. You just consider the particle number now, but no longer divide by the measure. How do you go from here to a *density* as the name of the plugin suggests? It's also no longer current with the comment above it.

By the way, typo: unaccurate -> inaccurate

gassmoeller replied:
Good point, I will update the comment. 

And yes this contradiction got me as well that is why I implemented it wrong the first time around. It is still using a particle density though. When you think about what happens during refinement it becomes more clear. The particle density (= particle number / volume) stays constant no matter how small the cells become and that is why in similar cases we usually scale things with the volume of the cell (as for example in the last lines of the 'density' refinement plugin that scales the density gradient with the cell volume). Now scaling the particle density with the volume of the cell would mean: (particle number / volume) * volume, and the volume cancels out. I can add this explanation to the comment as well.

gassmoeller commented:
I updated the documentation and fixed the test results (they were the old results before the patch, so I have proven that the test will catch this problem :smile:).
If the tester is happy and you agree with the documentation and my explanation, this should be ready to merge.

bangerth commented:
So, let me try to understand what you're trying to achieve. If I see this right, the goal of the postprocessor is to equilibrate the number of particles per cells, by refining those cells that have lots of particles, and coarsening those that have few. In that case, there is indeed no need for division by the cell's volume.

If this is correct, then maybe my confusion results from the naming of the plugin (both the class name and the name given to it in the plugin declaration at the bottom). Should it, in that case, rather be called "particle number", maybe? There is, after all, no density involved any more here :-)

gassmoeller commented:
I am not quite sure about the best name. 'particle number' would confuse me, because I would think: "Which number? Their id?". I would agree on "particle count", because that is also the name of the postprocessor that outputs exactly this quantity. However, I still think "particle density" is somehow the most self explaining name, independent of how it is implemented. I get a high resolution in regions with a high particle density and a low resolution in regions of low particle density. Both ways are fine with me though. Would you prefer 'particle count'?

bangerth commented:
Let's go with the name you have then, the explanation you now have for the plugin is the important part after all. Thanks for taking the time to explain it to me!
